lib/handler.js

```diff
@@ -16,40 +17,44 @@ exports.initTagDoc = function (event, callback) {
   // to get geonames tag/hierarchy we need to trigger the tag-e-geo lambda
   if (tag.location && tag.location.lat && tag.location.lon && tag._id.match(/hotel:mhid/)) {
     geonamesLambda(tag, function (err, data) {
-      if (!err && !data.FunctionError) {
-        var res = JSON.parse(data.Payload); // parse the result returned by Geo Lambda
-        var g = res[res.length - 1];        // get the last geo tag in the array
-        if (tag.tags && tag.tags.length > 0) { //
-          var geoTagIndex = -1; // no geo tag exists
-          tag.tags.forEach((t, index) => { // a Hotel tag should only have a single Geo Tag
-            // over-write the existin geonames tag
-            if (t.node.match(/geonames/)) {
-              geoTagIndex = index;
-            }
-          });
-        }
-        if (geoTagIndex > -1) { // over-write existing geo tag in tags array
-          tag.tags[geoTagIndex] = {
-            'node': g._id,
-            'edge': 'LOCATED_IN',
-            'displayName': g.displayName,
-            'source': 'geonames',
-            'active': true
-          };
-        } else { // create it from scratch
-          tag.tags.push({
-            'node': g._id,
-            'edge': 'LOCATED_IN',
-            'displayName': g.displayName,
-            'source': 'geonames',
-            'active': true
-          });
-        }
+      if (err || data.FunctionError) {
+        var err2 = err || new Error(data.FunctionError);
+        AwsHelper.log.info(err2);
+        return callback(err2, tag);
       }
-      return callback(err, tag);
+
+      var res = JSON.parse(data.Payload); // parse the result returned by Geo Lambda
+      var g = res[res.length - 1];        // get the last geo tag in the array
+      if (tag.tags && tag.tags.length > 0) { //
+        var geoTagIndex = -1; // no geo tag exists
+        tag.tags.forEach((t, index) => { // a Hotel tag should only have a single Geo Tag
+          // over-write the existin geonames tag
+          if (t.node.match(/geonames/)) {
+            geoTagIndex = index;
+          }
+        });
+      }
+      if (geoTagIndex > -1) { // over-write existing geo tag in tags array
+        tag.tags[geoTagIndex] = {
+          'node': g._id,
+          'edge': 'LOCATED_IN',
+          'displayName': g.displayName,
+          'source': 'geonames',
+          'active': true
+        };
+      } else { // create it from scratch
+        tag.tags.push({
+          'node': g._id,
+          'edge': 'LOCATED_IN',
+          'displayName': g.displayName,
+          'source': 'geonames',
+          'active': true
+        });
+      }
+      return callback(null, tag);
     });
   } else {
-    return callback(null, tag);
+    process.nextTick(callback.bind(null, null, tag));
```

Comment:
For those following along at home, @hassy explained this to me *verbally*:

instead of 
```js
return callback(null, data);
```
do:
```js
process.nextTick(callback.bind(null, null, data));
```
this is equivalent to:
```js
process.nextTick(function (err, data) { return callback(err, data); });
```
Because `callback.bind` returns a function and the first argument to `.bind` is `this` which we are setting to `null`.

The reason for the `process.nextTick` is to ensure that the function is always asynchronous.

see:
+ http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony
+ https://nodejs.org/api/process.html#process_process_nexttick_callback_arg